src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java

```diff
@@ -0,0 +1,424 @@
+                return;
+            }
+            InventoryUtils.addItemStackToInventory(chest, stack);
+            inventoryCitizen.removeStackFromSlot(i);
+        }
+
+    }
+
+    /**
+     * Search the right chest for an itemStack.
+     * @param stack the stack to dump.
+     * @return the tile entity of the chest
+     */
+    @Nullable
+    private TileEntityChest searchRightChestForStack(@NotNull final ItemStack stack)
+    {
+        if(InventoryUtils.findFirstSlotInInventoryWith(this, stack.getItem(), stack.getItemDamage()) != -1 && InventoryUtils.getOpenSlot(this) != -1)
+        {
+            return this;
+        }
+
+        for(@NotNull final BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            final TileEntity entity = worldObj.getTileEntity(pos);
+            if(entity instanceof TileEntityChest
+                    && InventoryUtils.findFirstSlotInInventoryWith((TileEntityChest) entity, stack.getItem(), stack.getItemDamage()) != -1
+                    && InventoryUtils.getOpenSlot(this) != -1)
+            {
+                return (TileEntityChest) entity;
+            }
+        }
+
+        @Nullable final TileEntityChest chest = searchChestWithSimilarItem(stack);
+        return chest == null ? searchMostEmptySlot() : chest;
+    }
+
+    /**
+     * Searches a chest with a similar item as the incoming stack.
+     * @param stack the stack.
+     * @return the entity of the chest.
+     */
+    @Nullable
+    private TileEntityChest searchChestWithSimilarItem(final ItemStack stack)
+    {
+        for(@NotNull final BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            final TileEntity entity = worldObj.getTileEntity(pos);
+            if(entity instanceof TileEntityChest
+                    && InventoryUtils.findFirstSlotInInventoryWith((TileEntityChest) entity, stack.getItem(), -1) != -1
+                    && InventoryUtils.getOpenSlot(this) != -1)
+            {
+                return (TileEntityChest) entity;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Search for the chest with the least items in it.
+     * @return the tileEntity of this chest.
+     */
+    @Nullable
+    private TileEntityChest searchMostEmptySlot()
+    {
+        int freeSlots = 0;
+        TileEntityChest emptiestChest = null;
+        for(@NotNull final BlockPos pos : getBuilding().getAdditionalCountainers())
+        {
+            final TileEntity entity = worldObj.getTileEntity(pos);
+            if(entity == null)
+            {
+                getBuilding().removeContainerPosition(pos);
+                continue;
+            }
+            if(entity instanceof TileEntityChest && InventoryUtils.getOpenSlot((IInventory) entity) != -1)
+            {
+                int tempFreeSlots = ((TileEntityChest) entity).getSizeInventory() - InventoryUtils.getAmountOfStacks((IInventory) entity);
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'tempFreeSlots' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)